File: engine/src/main/java/de/hanno/hpengine/engine/graphics/light/directional/DirectionalLight.java
package de.hanno.hpengine.engine.graphics.light.directional;

import de.hanno.hpengine.engine.backend.EngineContext;
import de.hanno.hpengine.engine.camera.Camera;
import de.hanno.hpengine.engine.component.InputControllerComponent;
import de.hanno.hpengine.engine.entity.Entity;
import de.hanno.hpengine.engine.graphics.shader.Program;
import org.jetbrains.annotations.NotNull;
import org.joml.Vector3f;

import static org.lwjgl.glfw.GLFW.*;

public class DirectionalLight extends Camera {

	private boolean castsShadows = false;
	private Vector3f color = new Vector3f(1,1,1);
	private float scatterFactor = 1f;

	@NotNull
	public Entity entity;

	public DirectionalLight(Entity entity) {
		super(entity);
		this.entity = entity;
		setPerspective(false);
		setColor(new Vector3f(1f, 0.76f, 0.49f));
		setScatterFactor(1f);
		setWidth(1500);
		setHeight(1500);
		setFar(-5000);
		entity.translation(new Vector3f(12f, 300f, 2f));
	}

	@Override
	public void update(float seconds) {
        super.update(seconds);
	}

	public void drawDebug(Program program) {
        throw new IllegalStateException("Currently not implemented!");
//		box.getComponentOption(ModelComponent.class).ifPresent(modelComponent -> {
//			modelComponent.drawDebug(program, getTransform().getTransformationBuffer());
//		});
	}

	public Vector3f getDirection () {
		return entity.getViewDirection();
	}

	public Vector3f getColor() {
		return color ;
	}
	public void setColor(Vector3f color) {
		this.color = color;
	}

	public float getScatterFactor() {
		return scatterFactor;
	}
	public void setScatterFactor(float scatterFactor) {
		this.scatterFactor = scatterFactor;
	}


	public void translate(Vector3f offset) {
		entity.translate(offset);
	}

	public static class DirectionalLightController extends InputControllerComponent {
		private static final long serialVersionUID = 1L;
		private EngineContext engine;

		public DirectionalLightController(EngineContext engine, @NotNull Entity entity) {
			super(entity);
			this.engine = engine;
		}

		@Override
		public void update(float seconds) {

			float moveAmount = 100 * seconds;
			float degreesPerSecond = 45;
			float rotateAmount = (float) Math.toRadians(degreesPerSecond) * seconds;

			if (engine.getInput().isKeyPressed(GLFW_KEY_UP)) {
				getEntity().rotateAround(new Vector3f(0, 1, 0), rotateAmount, new Vector3f());
			}
			if (engine.getInput().isKeyPressed(GLFW_KEY_DOWN)) {
				getEntity().rotateAround(new Vector3f(0, 1, 0), -rotateAmount, new Vector3f());
			}
			if (engine.getInput().isKeyPressed(GLFW_KEY_LEFT)) {
				getEntity().rotateAround(new Vector3f(1, 0, 0), rotateAmount, new Vector3f());
			}
			if (engine.getInput().isKeyPressed(GLFW_KEY_RIGHT)) {
				getEntity().rotateAround(new Vector3f(1, 0, 0), -rotateAmount, new Vector3f());
			}
			if (engine.getInput().isKeyPressed(GLFW_KEY_8)) {
				getEntity().translate(new Vector3f(0, -moveAmount, 0));
			}
			if (engine.getInput().isKeyPressed(GLFW_KEY_2)) {
				getEntity().translate(new Vector3f(0, moveAmount, 0));
			}
			if (engine.getInput().isKeyPressed(GLFW_KEY_4)) {
				getEntity().translate(new Vector3f(-moveAmount, 0, 0));
			}
			if (engine.getInput().isKeyPressed(GLFW_KEY_6)) {
				getEntity().translate(new Vector3f(moveAmount, 0, 0));
			}
		}
	}
}

File: engine/src/main/java/de/hanno/hpengine/engine/graphics/light/directional/DirectionalLight.kt
package de.hanno.hpengine.engine.graphics.light.directional

import com.artemis.*
import com.artemis.annotations.All
import com.artemis.annotations.Wire
import com.artemis.utils.Bag
import de.hanno.hpengine.engine.WorldPopulator
import de.hanno.hpengine.engine.component.artemis.DirectionalLightComponent
import de.hanno.hpengine.engine.component.artemis.NameComponent
import de.hanno.hpengine.engine.component.artemis.TransformComponent
import de.hanno.hpengine.engine.component.artemis.forEachEntity
import de.hanno.hpengine.engine.entity.CycleSystem
import de.hanno.hpengine.engine.graphics.state.RenderState
import de.hanno.hpengine.engine.input.Input
import de.hanno.hpengine.engine.system.Extractor
import de.hanno.hpengine.engine.transform.Transform
import org.jetbrains.kotlin.utils.addToStdlib.firstIsInstance
import org.jetbrains.kotlin.utils.addToStdlib.firstIsInstanceOrNull
import org.joml.AxisAngle4f
import org.joml.Matrix4f
import org.joml.Quaternionf
import org.joml.Vector3f
import org.lwjgl.glfw.GLFW


@All(DirectionalLightComponent::class, TransformComponent::class)
class DirectionalLightSystem: BaseEntitySystem(), Extractor, WorldPopulator {
    @Wire
    lateinit var input: Input
    lateinit var cycleSystem: CycleSystem
    lateinit var directionalLightComponentMapper: ComponentMapper<DirectionalLightComponent>
    lateinit var transformComponentMapper: ComponentMapper<TransformComponent>

    override fun processSystem() {
        val moveAmount = 100 * world.delta
        val degreesPerSecond = 45f
        val rotateAmount = Math.toRadians(degreesPerSecond.toDouble()).toFloat() * world.delta

        forEachEntity {
            val transform = transformComponentMapper[it].transform

            if (input.isKeyPressed(GLFW.GLFW_KEY_UP)) {
                transform.rotateAround(Vector3f(0f, 1f, 0f), rotateAmount, Vector3f())
            }
            if (input.isKeyPressed(GLFW.GLFW_KEY_DOWN)) {
                transform.rotateAround(Vector3f(0f, 1f, 0f), -rotateAmount, Vector3f())
            }
            if (input.isKeyPressed(GLFW.GLFW_KEY_LEFT)) {
                transform.rotateAround(Vector3f(1f, 0f, 0f), rotateAmount, Vector3f())
            }
            if (input.isKeyPressed(GLFW.GLFW_KEY_RIGHT)) {
                transform.rotateAround(Vector3f(1f, 0f, 0f), -rotateAmount, Vector3f())
            }
            if (input.isKeyPressed(GLFW.GLFW_KEY_8)) {
                transform.translate(Vector3f(0f, -moveAmount, 0f))
            }
            if (input.isKeyPressed(GLFW.GLFW_KEY_2)) {
                transform.translate(Vector3f(0f, moveAmount, 0f))
            }
            if (input.isKeyPressed(GLFW.GLFW_KEY_4)) {
                transform.translate(Vector3f(-moveAmount, 0f, 0f))
            }
            if (input.isKeyPressed(GLFW.GLFW_KEY_6)) {
                transform.translate(Vector3f(moveAmount, 0f, 0f))
            }
        }
    }

    override fun extract(currentWriteState: RenderState) {
        currentWriteState.directionalLightHasMovedInCycle = currentWriteState.cycle
        val entries: Map<Int, Bag<Component>> = currentWriteState.componentsForEntities.filterValues { it.firstIsInstanceOrNull<DirectionalLightComponent>() != null }
        if(entries.isEmpty()) return

        val light = entries.entries.first().value.firstIsInstance<DirectionalLightComponent>()
        val transform = entries.entries.first().value.firstIsInstance<TransformComponent>().transform

        val directionalLightState = currentWriteState.directionalLightState[0]

        directionalLightState.color.set(light.color)
        directionalLightState.direction.set(transform.viewDirection)
        directionalLightState.scatterFactor = light.scatterFactor
        val viewMatrix = Matrix4f(transform).invert()
        directionalLightState.viewMatrix.set(viewMatrix)
        directionalLightState.projectionMatrix.set(light.camera.projectionMatrix)
        directionalLightState.viewProjectionMatrix.set(Matrix4f(light.camera.projectionMatrix).mul(viewMatrix))
    }

    override fun World.populate() {
        addDirectionalLight()
    }
}


fun World.addDirectionalLight() {
    edit(create()).apply {
        create(NameComponent::class.java).apply {
            name = "DirectionalLight"
        }

        create(DirectionalLightComponent::class.java).apply { }
        create(TransformComponent::class.java).apply {
            transform = Transform().apply {
                translate(Vector3f(12f, 300f, 2f))
                rotateAroundLocal(Quaternionf(AxisAngle4f(Math.toRadians(100.0).toFloat(), 1f, 0f, 0f)), 0f, 0f, 0f)
            }
        }
    }
}
